### PR TITLE
Add Path.cs performance tests.

### PR DESCRIPTION
**Pull request by stephenmichaelf**
Add performance tests for Path.cs

This is based on #18765

@mellinoe @karelz 

I had two questions:

1. Why do the perf tests run with 9 calls to the method inside the loop?
2. How do you decide the number of inner iterations? AKA doing 20k vs 10k and 20k and 30k?

I did not add tests for:

AltDirectorySeparatorChar
DirectorySeparatorChar
InvalidPathChars
PathSeparator
VolumeSeparatorChar
GetInvalidFileNameChars
GetInvalidPathChars
GetTempFileName
GetRelativePath

### CODE REVIEW

**Comment by karelz** [comment]
BTW: @mellinoe is OOF until Tuesday I think.

@ianhays @JeremyKuhne can you please take a look in the meantime? Or should we wait for @mellinoe?

**Comment by stephenmichaelf** [comment]
@karelz Thanks for letting me know.

For whoever replies, if you could let me know how to debug the failure on CentOS7.1 I would appreciate it. Not sure if I can remote into the build machines.

**Comment by karelz** [comment]
It could be flaky test:
```
Expecting Child Task final Status to be Canceled, while getting RanToCompletion\nExpected: True\nActual: False
+++++++++++++++++++
STACK TRACE:
at System.Threading.Tasks.Tests.Status.TaskStatusTest.RealRun() in /mnt/resource/j/workspace/dotnet_corefx/master/centos7.1_release_prtest/src/System.Threading.Tasks/tests/Task/TaskStatusTest.cs:line
```

Search issues for the test name. If it exists, just rerun the CI leg (@dotnetbot test ...)

**Comment by karelz** [comment]
You can't remote into test machines. Even we can't do it well.

**Comment by danmosemsft** [comment]
@stephenmichaelf this may answer some questions https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/performance-tests.md

File: src/System.Runtime.Extensions/tests/Performance/Perf.Path.cs [hunk 1]
```diff
@@ -62,5 +62,169 @@ public void GetDirectoryName()
...
+        [InlineData(30000)]
+        public void GetFullPath(int innerIterations)
+        {
+            PerfUtils utils = new PerfUtils();
+            string testPath = utils.GetTestFilePath();
+            foreach (var iteration in Benchmark.Iterations)
+                using (iteration.StartMeasurement())
+                    for (int i = 0; i < innerIterations; i++)
+                    {
+                        Path.GetFullPath(testPath); Path.GetFullPath(testPath); Path.GetFullPath(testPath);
```
Comment (ianhays) [inline src/System.Runtime.Extensions/tests/Performance/Perf.Path.cs]: Doesn't GetFullPath operate as O(N) based on the length of the input path?

**Comment by ianhays** [comment]
>Why do the perf tests run with 9 calls to the method inside the loop?

This is usually done when the call itself completes quickly. We make the explicit call more than 1 time so that the overhead from the loop is lessened and we get more interesting results that are based on the performance of the thing being tested and not the loop infrastructure. 

>How do you decide the number of inner iterations? AKA doing 20k vs 10k and 20k and 30k?

It should depend on how long the individual calls take to complete. You want the total elapsed time to be high enough that machine interference is minimized. For example, if you were to do just 10 iterations then the results would be highly unreliable because the tiniest machine interference would screw the results such that they were worthless.

File: src/System.Runtime.Extensions/tests/Performance/Perf.Path.cs (hunk 1 above)
Comment (ianhays) [inline src/System.Runtime.Extensions/tests/Performance/Perf.Path.cs]: If so, then the 10k/20k/30k runs will be based on more than just the number of iterations but also the length of the `testPath`. That could skew the results.

Comment (JeremyKuhne) [inline src/System.Runtime.Extensions/tests/Performance/Perf.Path.cs]: It is radically different for path lengths. We should be testing paths that are < 250,  == 500, and == 1000 characters to keep track of perf (and work on all systems, as some top out at 1024). Paths don't have to physically exist on the system (and shouldn't for the purposes of path manipulation tests).

Comment (JeremyKuhne) [inline src/System.Runtime.Extensions/tests/Performance/Perf.Path.cs]: Those cases are "legacy length", "typical long path", "really long path". On some Unix variants you can go to 4K. Windows goes to 32K. Our assertion has been that we don't really care about > 1K for perf- particularly if it impacts < 1K.

Comment (JeremyKuhne) [inline src/System.Runtime.Extensions/tests/Performance/Perf.Path.cs]: > That could skew the results.

Yep, for all of these. They are very much bound to the length of the path- of which GetFullPath ends up playing a part in most of the APIs as most call GFP.

**Comment by stephenmichaelf** [comment]
@dotnet-bot test Innerloop CentOS7.1 Release x64 Build and Test

File: src/System.Runtime.Extensions/tests/Performance/Perf.Path.cs (hunk 1 above)
Comment (stephenmichaelf) [inline src/System.Runtime.Extensions/tests/Performance/Perf.Path.cs]: Thanks for the feedback @ianhays @JeremyKuhne I will make these changes!

**Comment by stephenmichaelf** [comment]
@ianhays Thank you, that is very helpful!

**Comment by stephenmichaelf** [comment]
@JeremyKuhne

I added methods for: GetFullPathForLegacyLength, GetFullPathForTypicalLongPath, and GetFullPathForReallyLongPath using the lengths you mentioned.

The latter two are throwing an exception(I am on Win 10):

```
[5/6/2017 4:17:03 PM][INF] System.IO.Tests.Perf_Path.GetFullPathForTypicalLongPath(innerIterations: 30000)
[5/6/2017 4:17:03 PM][ERR] System.IO.Tests.Perf_Path.GetFullPathForTypicalLongPath(innerIterations: 30000): 
The specified file name or path is too long, or a component of the specified path is too long.
D:\OpenSource\second-try\corefx\bin\Windows_NT.AnyCPU.Release\System.Runtime.Extensions.Performance.Tests\netcoreapp\PA4mJOKyEn1kNsBGRqVZLYqm5RXz2uQe4K8setZfzaDERV6KJ8Mnv5ZmpmyVJjGViN7bVCnn\GEpB0MyhpemBWkfgQFGzwXGfiV+jTR3ibl9l\L4hHh23En8YCCA9Dn\v6EDdaQvTrnYFNYEaG4tsX2Jte+kroHwqKS4Sm7lRonrf82SdBMVTkg5H2+icnTazq+i9j9KWhEfEs0AdwNK885NGOmgmjxqdGema44STjYw3Xy+Q09JoSS8c2JGFNZF7iyhVVqKOjNyN4HYo++CDcOWz5qHjK9+A3bE4yfaJF+QYXxicPnbpaJITAvNcP01OWWfr0U7vWJ8IlHjxVYMmBMiSEJzH5zKuyn3Pz+jsJJ5j6nBO3UaVKMZ5k2n+OBMO7T4qyi6whrJ1PQbQ34j2HG\ajasS12eCPOfWTlOIsJJqwOLqgVqvs2P+GaA\hEk9g3b6I6pn\BXWqM1G2jgerEx0pd6mC5sVKtoHFL2HQniemIscjo0nwwSM1BiNKJlr320Tyt5ql7fb5PiQYKdYD2ZU6AFswd3\DKLZ2TdCiAIA23VQs3eaNAETQOmUV6JddISyOEBX\t2PFL47UL5AWVz2T0VyB5vuurQAGYE0KrUAfKRy0E+BqNxXwV75XqOC0ueM3ggdKy0og72MV6UcWunI+w=
[5/6/2017 4:17:03 PM][ERR] 
	 at System.IO.PathHelper.Normalize(String path, Boolean checkInvalidCharacters, Boolean expandShortPaths)
     at System.IO.Path.GetFullPath(String path)
     at System.IO.Tests.Perf_Path.GetFullPathForTypicalLongPath(Int32 innerIterations) in D:\OpenSource\second-try\corefx\src\System.Runtime.Extensions\tests\Performance\Perf.Path.cs:line 152
```

For the example above the string I pass in to GetFullPath is 500 random characters and the total path we see in the exception is 768.

I am a bit confused. Is the older limit of 250 still taking effect? I searched the directory for PathHelper to find the Normalize method and debug the issue that way and all I can find is PathHelpers.

What am I misunderstanding?

**Comment by JeremyKuhne** [comment]
> I am a bit confused. Is the older limit of 250 still taking effect? I searched the directory for PathHelper to find the Normalize method and debug the issue that way and all I can find is PathHelpers.

You still can't have path segments that are over 255 characters. Make sure you have a directory separator at least that often.

File: src/Common/tests/System/PerfUtils.cs [hunk 2]
```diff
@@ -31,14 +32,27 @@ public PerfUtils(int seed)
...
-            _rand.NextBytes(bytes);
-            return Convert.ToBase64String(bytes);
+            char[] str = new char[length];
+
+            for (int i = 0; i < str.Length; i++)
+            {
+                // Add path separator so folders aren't too long.
+                if (i%20 == 0)
+                {
+                    str[i] = '\\';
```
Comment (JeremyKuhne) [inline src/Common/tests/System/PerfUtils.cs]: Use Path.DirectorySeparator to make sure you're good cross plat

**Comment by stephenmichaelf** [comment]
@JeremyKuhne Thank you for the help and patience! :)

**Comment by stephenmichaelf** [comment]
@karelz Could you merge this?

**Comment by karelz** [comment]
@JeremyKuhne can you please merge if you are happy with it?

**Pull request merged**